
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>References &#8212; ehrapy 0.2.0 documentation</title>
    
    <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
  
    
    <link rel="stylesheet"
      href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
    <link rel="preload" as="font" type="font/woff2" crossorigin
      href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
    <link rel="preload" as="font" type="font/woff2" crossorigin
      href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">
  
    
      
  
    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/pydata-sphinx-theme.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    <link rel="stylesheet" type="text/css" href="_static/custom_cookietemple.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx_gallery.css" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx.css" />
    <link rel="stylesheet" type="text/css" href="_static/dataframe.css" />
    
    <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">
  
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Credits" href="authors.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    
    <nav class="navbar navbar-light navbar-expand-lg bg-light fixed-top bd-navbar" id="navbar-main"><div class="container-xl">

  <div id="navbar-start">
    
    

<a class="navbar-brand" href="index.html">
  <img src="_static/placeholder.png" class="logo" alt="logo">
</a>


    
  </div>

  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-collapsible" aria-controls="navbar-collapsible" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  
  <div id="navbar-collapsible" class="col-lg-9 collapse navbar-collapse">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <ul id="navbar-main-elements" class="navbar-nav">
    <li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="installation.html">
  Installation
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="usage/usage.html">
  Usage
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="tutorials/index.html">
  Tutorials
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="contributing.html">
  Contributor Guide
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="code_of_conduct.html">
  Code of Conduct
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference external nav-link" href="https://github.com/theislab/ehrapy/discussions">
  Discussions
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="authors.html">
  Credits
 </a>
</li>

<li class="toctree-l1 current active nav-item">
 <a class="current reference internal nav-link" href="#">
  References
 </a>
</li>

    
</ul>
      </div>
      
    </div>

    <div id="navbar-end">
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/theislab/ehrapy" rel="noopener" target="_blank" title="GitHub"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="https://twitter.com/lukasheumos" rel="noopener" target="_blank" title="Twitter"><span><i class="fab fa-twitter-square"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>
</div>
    </nav>
    

    <div class="container-xl">
      <div class="row">
          
            
            <!-- Only show if we have sidebars configured, else just a small margin  -->
            <div class="col-12 col-md-3 bd-sidebar">
              <div class="sidebar-start-items"><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <div class="bd-toc-item active">
    
  </div>
</nav>
              </div>
              <div class="sidebar-end-items">
              </div>
            </div>
            
          

          
          <div class="d-none d-xl-block col-xl-2 bd-toc">
            
              
              <div class="toc-item">
                

<nav id="bd-toc-nav">
    
</nav>
              </div>
              
              <div class="toc-item">
                

<div class="tocsection editthispage">
    <a href="https://github.com/theislab/ehrapy/edit/development/docs/references.rst">
        <i class="fas fa-pencil-alt"></i> Edit this page
    </a>
</div>

              </div>
              
            
          </div>
          

          
          
            
          
          <main class="col-12 col-md-9 col-xl-7 py-md-5 pl-md-5 pr-md-4 bd-content" role="main">
              
              <div>
                
  <div class="section" id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<dl class="citation">
<dt class="label" id="amir13"><span class="brackets">Amir13</span></dt>
<dd><p>Amir <em>et al.</em> (2013),
<em>viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia</em>,
<a class="reference external" href="https://doi.org/10.1038/nbt.2594">Nature Biotechnology</a>.</p>
</dd>
<dt class="label" id="blondel08"><span class="brackets">Blondel08</span></dt>
<dd><p>Blondel <em>et al.</em> (2008),
<em>Fast unfolding of communities in large networks</em>,
<a class="reference external" href="https://doi.org/10.1088/1742-5468/2008/10/P10008">J. Stat. Mech.</a>.</p>
</dd>
<dt class="label" id="chippada18"><span class="brackets">Chippada18</span></dt>
<dd><p><em>ForceAtlas2 for Python and NetworkX</em>,
<a class="reference external" href="https://github.com/bhargavchippada/forceatlas2">GitHub</a>.</p>
</dd>
<dt class="label" id="coifman05"><span class="brackets">Coifman05</span></dt>
<dd><p>Coifman <em>et al.</em> (2005),
<em>Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps</em>,
<a class="reference external" href="https://doi.org/10.1038/nmeth.3971">PNAS</a>.</p>
</dd>
<dt class="label" id="csardi06"><span class="brackets">Csardi06</span></dt>
<dd><p>Csardi <em>et al.</em> (2006),
<em>The igraph software package for complex network research</em>,
<a class="reference external" href="http://igraph.org">InterJournal Complex Systems</a>.</p>
</dd>
<dt class="label" id="fruchterman91"><span class="brackets">Fruchterman91</span></dt>
<dd><p>Fruchterman &amp; Reingold (1991),
<em>Graph drawing by force-directed placement</em>,
<a class="reference external" href="http://doi.org:10.1002/spe.4380211102">Software: Practice &amp; Experience</a>.</p>
</dd>
<dt class="label" id="haghverdi15"><span class="brackets">Haghverdi15</span></dt>
<dd><p>Haghverdi <em>et al.</em> (2015),
<em>Diffusion maps for high-dimensional single-cell analysis of differentiation data</em>,
<a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv325">Bioinformatics</a>.</p>
</dd>
<dt class="label" id="haghverdi16"><span class="brackets">Haghverdi16</span></dt>
<dd><p>Haghverdi <em>et al.</em> (2016),
<em>Diffusion pseudotime robustly reconstructs branching cellular lineages</em>,
<a class="reference external" href="https://doi.org/10.1038/nmeth.3971">Nature Methods</a>.</p>
</dd>
<dt class="label" id="islam11"><span class="brackets">Islam11</span></dt>
<dd><p>Islam <em>et al.</em> (2011),
<em>Characterization of the single-cell transcriptional landscape by highly multiplex RNA-seq</em>,
<a class="reference external" href="https://doi.org/10.1101/gr.110882.110">Genome Research</a>.</p>
</dd>
<dt class="label" id="jacomy14"><span class="brackets">Jacomy14</span></dt>
<dd><p>Jacomy <em>et al.</em> (2014),
<em>ForceAtlas2, a Continuous Graph Layout Algorithm for Handy Network Visualization Designed for the Gephi Software</em>
<a class="reference external" href="https://doi.org/10.1371/journal.pone.0098679">PLOS One</a>.</p>
</dd>
<dt class="label" id="johnson07"><span class="brackets">Johnson07</span></dt>
<dd><p>Johnson, Li &amp; Rabinovic (2007),
<em>Adjusting batch effects in microarray expression data using empirical Bayes methods</em>,
<a class="reference external" href="https://doi.org/10.1093/biostatistics/kxj037">Biostatistics</a>.</p>
</dd>
<dt class="label" id="lambiotte09"><span class="brackets">Lambiotte09</span></dt>
<dd><p>Lambiotte <em>et al.</em> (2009)
<em>Laplacian Dynamics and Multiscale Modular Structure in Networks</em>
<a class="reference external" href="https://arxiv.org/abs/0812.1770">arXiv</a>.</p>
</dd>
<dt class="label" id="leek12"><span class="brackets">Leek12</span></dt>
<dd><p>Leek <em>et al.</em> (2012),
<em>sva: Surrogate Variable Analysis. R package</em>
<a class="reference external" href="https://doi.org/10.18129/B9.bioc.sva">Bioconductor</a>.</p>
</dd>
<dt class="label" id="levine15"><span class="brackets">Levine15</span></dt>
<dd><p>Levine <em>et al.</em> (2015),
<em>Data-Driven Phenotypic Dissection of AML Reveals Progenitor–like Cells that Correlate with Prognosis</em>,
<a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.047">Cell</a>.</p>
</dd>
<dt class="label" id="maaten08"><span class="brackets">Maaten08</span></dt>
<dd><p>Maaten &amp; Hinton (2008),
<em>Visualizing data using t-SNE</em>,
<a class="reference external" href="http://www.jmlr.org/papers/v9/vandermaaten08a.html">JMLR</a>.</p>
</dd>
<dt class="label" id="mcinnes18"><span class="brackets">McInnes18</span></dt>
<dd><p>McInnes &amp; Healy (2018),
<em>UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction</em>,
<a class="reference external" href="https://arxiv.org/abs/1802.03426">arXiv</a>.</p>
</dd>
<dt class="label" id="pedersen12"><span class="brackets">Pedersen12</span></dt>
<dd><p>Pedersen (2012),
<em>Python implementation of ComBat</em>
<a class="reference external" href="https://github.com/brentp/combat.py">GitHub</a>.</p>
</dd>
<dt class="label" id="pedregosa11"><span class="brackets">Pedregosa11</span></dt>
<dd><p>Pedregosa <em>et al.</em> (2011),
<em>Scikit-learn: Machine Learning in Python</em>,
<a class="reference external" href="http://www.jmlr.org/papers/v12/pedregosa11a.html">JMLR</a>.</p>
</dd>
<dt class="label" id="traag17"><span class="brackets">Traag17</span></dt>
<dd><p>(2017),
<em>Louvain</em>,
<a class="reference external" href="https://doi.org/10.5281/zenodo.35117">GitHub</a>.</p>
</dd>
<dt class="label" id="traag18"><span class="brackets">Traag18</span></dt>
<dd><p>Traag <em>et al.</em> (2018),
<em>From Louvain to Leiden: guaranteeing well-connected communities</em>
<a class="reference external" href="https://arxiv.org/abs/1810.08473">arXiv</a>.</p>
</dd>
<dt class="label" id="weinreb17"><span class="brackets">Weinreb17</span></dt>
<dd><p>Weinreb <em>et al.</em> (2016),
<em>SPRING: a kinetic interface for visualizing high dimensional single-cell expression data</em>,
<a class="reference external" href="https://doi.org/10.1101/090332">bioRxiv</a>.</p>
</dd>
<dt class="label" id="wolf18"><span class="brackets">Wolf18</span></dt>
<dd><p>Wolf <em>et al.</em> (2018),
<em>Scanpy: large-scale single-cell gene expression data analysis</em>,
<a class="reference external" href="https://doi.org/10.1186/s13059-017-1382-0">Genome Biology</a>.</p>
</dd>
<dt class="label" id="wolf19"><span class="brackets">Wolf19</span></dt>
<dd><p>Wolf <em>et al.</em> (2019),
<em>PAGA: Graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells.</em>
<a class="reference external" href="https://doi.org/10.1186/s13059-019-1663-x">Genome Biology</a>,
<a class="reference external" href="https://doi.org/10.1101/208819">bioRxiv</a>.</p>
</dd>
<dt class="label" id="zunder15"><span class="brackets">Zunder15</span></dt>
<dd><p>Zunder <em>et al.</em> (2015),
<em>A continuous molecular roadmap to iPSC reprogramming through progression analysis of single-cell mass cytometry</em>,
<a class="reference external" href="https://doi.org/10.1016/j.stem.2015.01.015">Cell Stem Cell</a>.</p>
</dd>
</dl>
</div>


              </div>
              
              
              <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="authors.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Credits</p>
        </div>
    </a>
</div>
              
          </main>
          

      </div>
    </div>
  
    <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>
  <footer class="footer mt-5 mt-md-0">
  <div class="container">
    
    <div class="footer-item">
      <p class="copyright">
    &copy; Copyright 2021, Lukas Heumos, Theislab.<br>
</p>
    </div>
    
    <div class="footer-item">
      <p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 4.4.0.<br>
</p>
    </div>
    
  </div>
</footer>
  </body>
</html>